src/main/java/org/broadinstitute/hellbender/utils/downsampling/ReservoirDownsampler.java

```diff
@@ -110,35 +120,39 @@ public void submit ( final GATKRead newRead ) {
 
     @Override
     public boolean hasFinalizedItems() {
-        return ! reservoir.isEmpty();
+        // All items in the reservoir are pending until endOfInputStream is seen, at which point all items become finalized
+        return endOfInputStream && !reservoir.isEmpty();
     }
 
     @Override
     public List<GATKRead> consumeFinalizedItems() {
+        Utils.validate(endOfInputStream == true, "signalEndOfInput must be called before finalized items can be consumed");
         if (hasFinalizedItems()) {
             // pass reservoir by reference rather than make a copy, for speed
             final List<GATKRead> downsampledItems = reservoir;
             clearItems();
             return downsampledItems;
         } else {
             // if there's nothing here, don't bother allocating a new list
+            clearItems();
```

Comment:
See above -- calling `consumeFinalizedItems()` when there are no finalized items should return an empty list, and not change the internal state of the downsampler (eg., by calling `clearItems()`).

Reply:
But that would require the caller to detect when `consumeFinalizedItems` returns 0 items, and then call `clearItems()`, in order to reset the state. Doesn't that unnecessarily complicate the contract ?

Reply:
At some point we should consider adding a  method to the `Downsampler` interface called something like `resetState` to make the contract on how to recycle these more explicit.

Reply:
Most downsamplers don't get reused after `signalEndOfInput()` -- the `ReservoirDownsampler` within a downsampling iterator is an exception. It also doesn't make sense to me for `consumeFinalizedItems()` to have side effects like resetting the downsampler if there are no finalized items. Returning an empty list without any side effects seems much more logical.